templates.go

```diff
@@ -28,6 +28,126 @@ import (
 )
 `))
 
+type jsHeaderParams struct {
+	ToolName      string
+	FileName      string
+	ServiceName   string
+	Package       string
+	ManualImport string
+	TargetName    string
+	BuildTags     string
+	Methods       []jsRpcParams
+}
+
+type jsRpcParams struct {
+	ServiceName       string
+	MethodName        string
+	RequestType       string
+	ResponseStreaming bool
+	ApiPrefix         string
+	TargetName        string
+}
+
+var jsTemplate = template.Must(template.New("jsHeader").Funcs(funcMap).Parse(`// Code generated by {{.ToolName}}. DO NOT EDIT.
+// source: {{.FileName}}
+{{if .BuildTags}}
+{{.BuildTags}}
+{{end}}
+package {{.Package}}
+
+import (
+	"context"
+
+	gateway "github.com/grpc-ecosystem/grpc-gateway/runtime"
+{{- if .ManualImport }}
+	"{{.ManualImport}}"{{- end}}
+	"google.golang.org/grpc"
+)
+
+{{- define "unaryRpcFunc"}}
+		req := &{{.RequestType}}{}
+		err := marshaler.Unmarshal([]byte(reqJSON), req)
+		if err != nil {
+			callback("", err)
+			return
+		}
+
+		client := New{{.ServiceName}}Client(conn)
+		resp, err := client.{{.MethodName}}(ctx, req)
+		if err != nil {
+			callback("", err)
+			return
+		}
+
+		respBytes, err := marshaler.Marshal(resp)
+		if err != nil {
+			callback("", err)
+			return
+		}
+		callback(string(respBytes), nil)
+{{- end}}
+
+{{- define "streamRpcFunc"}}
+		req := &{{.RequestType}}{}
+		err := marshaler.Unmarshal([]byte(reqJSON), req)
+		if err != nil {
+			callback("", err)
+			return
+		}
+
+		client := New{{.ServiceName}}Client(conn)
+		stream, err := client.{{.MethodName}}(ctx, req)
+		if err != nil {
+			callback("", err)
+			return
+		}
+
+		go func() {
+			for {
+				select {
+				case <-stream.Context().Done():
+					callback("", stream.Context().Err())
+					return
+				default:
+				}
+
+				resp, err := stream.Recv()
+				if err != nil {
+					callback("", err)
+					return
+				}
+
+				respBytes, err := marshaler.Marshal(resp)
+				if err != nil {
+					callback("", err)
+					return
+				}
+				callback(string(respBytes), nil)
+			}
+		}()
+{{- end}}
+
+func Register{{.ServiceName | UpperCase}}JSONCallbacks(registry map[string]func(ctx context.Context,
```

Comment:
So the idea here is to autogenerate code to allow communication between the Go <-> JS layers, right? Might be useful to start a `docs` folder here with some high level details about how everything slots into a final app/system.

Reply:
Yes, that's the idea. I added a new section to the main README for now.